File: cpp/dijkstra.cpp
struct Edge{
    int to;
    long long cost;
    Edge(){}
    Edge(int to, long long cost):to(to), cost(cost){}
};

vector<long long> dijkstra(int start, const  vector<vector<Edge>>& es){
    int n = es.size();
    vector<long long> dist(n, INF);
    priority_queue<pair<long long,int>, vector<pair<long long,int>>, greater<pair<long long,int>>> que;
    dist[start] = 0;
    que.push({0, start});
    while(!que.empty()){
        long long w = que.top().first;
        int v = que.top().second;
        que.pop();

        if(dist[v] < w)continue;

        int m = es[v].size();
        for(int i = 0;i < m; ++i){
            const Edge& e = es[v][i];
            if(dist[e.to] > dist[v] + e.cost){
                dist[e.to] = dist[v] + e.cost;
                que.push({dist[e.to], e.to});
            }
        }
    }
    return dist;
}

File: rust/dijkstra.rs
#[derive(Clone)]
struct Edge{
    to: usize,
    cost: i64,
}

fn dijkstra(start: usize, es: &Vec<Vec<Edge>>) -> Vec<i64>{
    let n = es.len();
    let mut dist = vec![INF; n];
    let mut que = std::collections::BinaryHeap::new();

    dist[start] = 0;
    que.push((-0, start));
    while let Some((mut w, v)) = que.pop() {
        w *= -1;
        if dist[v] < w { continue; }

        for e in &es[v] {
            if dist[e.to] > dist[v] + e.cost {
                dist[e.to] = dist[v] + e.cost;
                que.push((-dist[e.to], e.to));
            }
        }
    }

    dist
}
